### ISSUE TITLE
ERROR while installing plugin

### ISSUE THREAD

**Issue by ostojicmilica** [y-ken/fluent-plugin-geoip#19]
Hi folks,

I can't install plugin, here is the error I receive:

```
 td-agent-gem install fluent-plugin-geoip

WARN: Unresolved specs during Gem::Specification.reset:
      json (>= 1.4.3)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
Building native extensions.  This could take a while...
ERROR:  Error installing fluent-plugin-geoip:
	ERROR: Failed to build gem native extension.

    /opt/td-agent/embedded/bin/ruby extconf.rb
checking for iconv_open() in iconv.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/td-agent/embedded/bin/ruby
	--with-geoip-dir
	--without-geoip-dir
	--with-geoip-include
	--without-geoip-include=${geoip-dir}/include
	--with-geoip-lib
	--without-geoip-lib=${geoip-dir}/lib
/opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:467:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:552:in `try_link0'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:567:in `try_link'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:753:in `try_func'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1038:in `block in have_func'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:929:in `block in checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block (2 levels) in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:347:in `postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:928:in `checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1037:in `have_func'
	from extconf.rb:8:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/geoip-c-0.9.1 for inspection.
Results logged to /opt/td-agent/embedded/lib/ruby/gems/2.1.0/extensions/x86_64-linux/2.1.0/geoip-c-0.9.1/gem_make.out
```

 Do you have idea what could be wrong?

**Comment by y-ken** [comment]
Hi @ostojicmilica 

Have you installed dependent library?

```sh
# for RHEL/CentOS
$ sudo yum install geoip-devel --enablerepo=epel

# for Ubuntu/Debian
$ sudo apt-get install libgeoip-dev
```

**Comment by ostojicmilica** [comment]
Yes I am.

Package GeoIP-devel-1.4.8-1.5.amzn1.x86_64 already installed and latest version
Nothing to do

**Comment by y-ken** [comment]
OK. Could you please trying to install `sudo yum zlib-devel`?

**Comment by ostojicmilica** [comment]
It is already installed.

2015-05-08 15:22 GMT+02:00 Y.Kentaro <notifications@github.com>:

> OK. Could you please trying to install sudo yum zlib-devel?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/y-ken/fluent-plugin-geoip/issues/19#issuecomment-100230899>
> .
>



-- 
Milica Ostojic

**Comment by y-ken** [comment]
Could you show me the logfile of `/opt/td-agent/embedded/lib/ruby/gems/2.1.0/extensions/x86_64-linux/2.1.0/geoip-c-0.9.1/gem_make.out`?
It may help us to resolve this issue!

**Comment by ostojicmilica** [comment]
checking for iconv_open() in iconv.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/td-agent/embedded/bin/ruby
	--with-geoip-dir
	--without-geoip-dir
	--with-geoip-include
	--without-geoip-include=${geoip-dir}/include
	--with-geoip-lib
	--without-geoip-lib=${geoip-dir}/lib
/opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:467:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:552:in `try_link0'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:567:in `try_link'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:753:in `try_func'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1038:in `block in have_func'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:929:in `block in checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block (2 levels) in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:347:in `postpone'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:928:in `checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1037:in `have_func'
	from extconf.rb:8:in `<main>'

extconf failed, exit code 1

**Comment by ostojicmilica** [comment]
yum group install "Development Tools" - solved this problem :)

Thanks a lot!

**Comment by y-ken** [comment]
I'll update the README to improbe more user friendly.
Thank you.

**Comment by y-ken** [comment]
README has updated. 
https://github.com/y-ken/fluent-plugin-geoip/commit/932fba4fa58d2b3731622250047f14ec444dbb50

**Comment by ostojicmilica** [comment]
Great! :)

**Comment by repeatedly** [comment]
This is good timing to use pure ruby implementation, geoip gem, by default.
If users need a performance, use geoip-c gem instead.

**Comment by y-ken** [comment]
@repeatedly 
Sure. I'm considering implement pure ruby version for the next release.

**Comment by gmpuran** [comment]
@repeatedly , getting this issue on Ubuntu 14.04, while trying to install plugin from Puppet. Have installed ruby-dev, libcurl4-gnutls-dev, libgeoip-dev and build-essential before installing plugin.

Also, have tried installation using td-agent-gem and fluent-gem command, both didn't work.

Please advise.

**Comment by y-ken** [comment]
Hi @gmpuran 

Would you show me the command that you have used to install this plugin?
And please tell me which version of td-agent used?

**Comment by gmpuran** [comment]
@y-ken , In puppet I'd mentioned in following two ways (inside exec):

command => "td-agent-gem install fluent-plugin-geoip",
    	path => [ '/usr/sbin' ],

and, 

command => "fluent-gem install fluent-plugin-geoip",
    	path => [ '/opt/td-agent/embedded/bin' ],

Version: 
# td-agent --version
td-agent 0.12.7

**Comment by y-ken** [comment]
Hi @gmpuran

OK. Are you using td-agent2?
How about execute with root like below?

```perl
exec { "td-agent-install-plugin-geoip":
  command => "td-agent-gem install fluent-plugin-geoip",
  path => [ '/usr/sbin' ],
  user => root
}
```

Another idea, turn it on verbose mode with `--verbose` for `puppet agent` command.

**Comment by gmpuran** [comment]
@y-ken , Yes, td-agent2 - have used following command for installation:

curl -L https://td-toolbelt.herokuapp.com/sh/install-ubuntu-trusty-td-agent2.sh | sh

No luck with executing as root, received same error.

This file: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/extensions/x86_64-linux/2.1.0/geoip-c-0.9.1/gem_make.out - also looks same as mentioned in above replies.

Not sure what it is not able to refer while running from puppet but able to refer when invoked manually.

**Comment by repeatedly** [comment]
Could you paste the content of `/opt/td-agent/embedded/lib/ruby/gems/2.1.0/extensions/x86_64-linux/2.1.0/geoip-c-0.9.1/gem_make.out` ?
Is this exactly same as ostojicmilica's one?

**Comment by gmpuran** [comment]
Yes, exactly same.. following is the content:

```javascript
/opt/td-agent/embedded/bin/ruby extconf.rb
checking for iconv_open() in iconv.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/opt/td-agent/embedded/bin/ruby
        --with-geoip-dir
        --without-geoip-dir
        --with-geoip-include
        --without-geoip-include=${geoip-dir}/include
        --with-geoip-lib
        --without-geoip-lib=${geoip-dir}/lib
/opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:467:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:552:in `try_link0'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:567:in `try_link'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:753:in `try_func'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1038:in `block in have_func'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:929:in `block in checking_for'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block (2 levels) in postpone'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:351:in `block in postpone'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:321:in `open'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:347:in `postpone'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:928:in `checking_for'
        from /opt/td-agent/embedded/lib/ruby/2.1.0/mkmf.rb:1037:in `have_func'
        from extconf.rb:8:in `<main>'

extconf failed, exit code 1
```

**Comment by gmpuran** [comment]
@y-ken , The error is solved. My bad.. path mentioned in exec is not sufficient. That is, when exec triggers the command, $PATH holds only "/usr/sbin", so other binaries are unknown (required to build native extensions).

Have configured full required path in exec, and the plugin is getting installed successfully.

**Comment by martinhynar** [comment]
For the record, I had this same problem while installing fluentd-plugin-route and 'gcc' was the only package I needed (+ its dependencies of course). This much less than whole 'Development Tools'.

**Comment by Matiasmct** [comment]
I know it's been few years, but I'm having the same issue. I have installed all dependencies but still no luck:

```
HEAD https://rubygems.org/api/v1/dependencies
200 OK
GET https://rubygems.org/api/v1/dependencies?gems=fluent-plugin-geoip
200 OK
GET https://rubygems.org/api/v1/dependencies?gems=dig_rb,fluentd,geoip-c,geoip2_c
200 OK
GET https://rubygems.org/api/v1/dependencies?gems=cool.io,http_parser.rb,msgpack,serverengine,sigdump,strptime,tzinfo,tzinfo-data,yajl-ruby
200 OK
GET https://rubygems.org/api/v1/dependencies?gems=concurrent-ruby
200 OK
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/.gitignore
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/Changelog.md
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/Gemfile
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/README.md
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/Rakefile
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/ext/geoip/extconf.rb
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/ext/geoip/geoip.c
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/geoip-c.gemspec
/opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/test/test_geoip.rb
Building native extensions. This could take a while...
current directory: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/ext/geoip
["/opt/td-agent/embedded/bin/ruby", "-I", "/opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0", "-r", "./siteconf20200722-1373-1preszg.rb", "extconf.rb"]
checking for iconv_open() in iconv.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/td-agent/embedded/bin/$(RUBY_BASE_NAME)
	--with-geoip-dir
	--without-geoip-dir
	--with-geoip-include
	--without-geoip-include=${geoip-dir}/include
	--with-geoip-lib
	--without-geoip-lib=${geoip-dir}/lib
/opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:468:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:553:in `try_link0'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:568:in `try_link'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:780:in `try_func'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:1067:in `block in have_func'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:957:in `block in checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:351:in `block (2 levels) in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:351:in `block in postpone'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:321:in `open'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:347:in `postpone'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:956:in `checking_for'
	from /opt/td-agent/embedded/lib/ruby/2.4.0/mkmf.rb:1066:in `have_func'
	from extconf.rb:8:in `<main>'
ERROR:  Error installing fluent-plugin-geoip:
	ERROR: Failed to build gem native extension.

    current directory: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1/ext/geoip
/opt/td-agent/embedded/bin/ruby -I /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0 -r ./siteconf20200722-1373-1preszg.rb extconf.rb
Building has failed. See above output for more information on the failure.
To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /opt/td-agent/embedded/lib/ruby/gems/2.4.0/extensions/x86_64-linux/2.4.0/geoip-c-0.9.1/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/geoip-c-0.9.1 for inspection.
Results logged to /opt/td-agent/embedded/lib/ruby/gems/2.4.0/extensions/x86_64-linux/2.4.0/geoip-c-0.9.1/gem_make.out
```

The content of the mkmf.log:

```
"gcc -o conftest -I/opt/td-agent/embedded/include/ruby-2.4.0/x86_64-linux -I/opt/td-agent/embedded/include/ruby-2.4.0/ruby/backward -I/opt/td-agent/embedded/include/ruby-2.4.0 -I. -I/opt/td-agent/embedded/include -O2 -D_FORTIFY_SOURCE=2 -fstack-protector -O3 -g -pipe -I/opt/td-agent/embedded/include   -I/opt/td-agent/embedded/include -O2 -D_FORTIFY_SOURCE=2 -fstack-protector -O3 -g -pipe -fPIC  conftest.c  -L. -L/opt/td-agent/embedded/lib -Wl,-rpath,/opt/td-agent/embedded/lib -L/opt/td-agent/embedded/lib -Wl,-rpath,/opt/td-agent/embedded/lib -L. -Wl,-rpath,/opt/td-agent/embedded/lib -L/opt/td-agent/embedded/lib -fstack-protector -rdynamic -Wl,-export-dynamic -L/opt/td-agent/embedded/lib  -Wl,-rpath,/opt/td-agent/embedded/lib      -Wl,-rpath,/opt/td-agent/embedded/lib -L/opt/td-agent/embedded/lib -lruby  -lpthread -lrt -ldl -lcrypt -lm   -lc"
In file included from /opt/td-agent/embedded/include/ruby-2.4.0/ruby.h:33,
                 from conftest.c:1:
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1381: error: wrong number of arguments specified for ‘__deprecated__’ attribute
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1390: error: wrong number of arguments specified for ‘__deprecated__’ attribute
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h: In function ‘rb_mul_size_overflow’:
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘c2’
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: ‘c2’ undeclared (first use in this function)
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: (Each undeclared identifier is reported only once
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: for each function it appears in.)
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: expected ‘)’ before ‘__int128’
/opt/td-agent/embedded/include/ruby-2.4.0/ruby/ruby.h:1615: error: expected ‘)’ before ‘__int128’
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return 0;
6: }
/* end */
```

Really struggling with this.

Thanks,
